tracer/src/Datadog.Trace/Configuration/ExporterSettings.cs

```diff
@@ -70,49 +75,63 @@ public ExporterSettings()
 
             ValidationWarnings = new List<string>();
 
             source ??= NullConfigurationSource.Instance;
 
             // Get values from the config
-            var traceAgentUrl = source.GetString(ConfigurationKeys.AgentUri);
-            var tracesPipeName = source.GetString(ConfigurationKeys.TracesPipeName);
-            var tracesUnixDomainSocketPath = source.GetString(ConfigurationKeys.TracesUnixDomainSocketPath);
+            var traceAgentUrl = new ConfigurationBuilder(source, telemetry).WithKeys(ConfigurationKeys.AgentUri).AsString().Get();
+            var tracesPipeName = new ConfigurationBuilder(source, telemetry).WithKeys(ConfigurationKeys.TracesPipeName).AsString().Get();
+            var tracesUnixDomainSocketPath = new ConfigurationBuilder(source, telemetry).WithKeys(ConfigurationKeys.TracesUnixDomainSocketPath).AsString().Get();
 
-            var tracesPipeTimeoutMs = source.GetInt32(ConfigurationKeys.TracesPipeTimeoutMs) ?? 0;
-            var agentHost = source.GetString(ConfigurationKeys.AgentHost) ??
-                        // backwards compatibility for names used in the past
-                        source.GetString("DD_TRACE_AGENT_HOSTNAME") ??
-                        source.GetString("DATADOG_TRACE_AGENT_HOSTNAME");
+            var agentHost = new ConfigurationBuilder(source, telemetry)
+                           .WithKeys(ConfigurationKeys.AgentHost, "DD_TRACE_AGENT_HOSTNAME", "DATADOG_TRACE_AGENT_HOSTNAME")
+                           .AsString()
+                           .Get();
 
-            var agentPort = source.GetInt32(ConfigurationKeys.AgentPort) ??
-                        // backwards compatibility for names used in the past
-                        source.GetInt32("DATADOG_TRACE_AGENT_PORT");
+            var agentPort = new ConfigurationBuilder(source, telemetry)
+                           .WithKeys(ConfigurationKeys.AgentPort, "DATADOG_TRACE_AGENT_PORT")
+                           .AsInt32()
+                           .Get();
 
-            var dogStatsdPort = source.GetInt32(ConfigurationKeys.DogStatsdPort) ?? 0;
-            var metricsPipeName = source.GetString(ConfigurationKeys.MetricsPipeName);
-            var metricsUnixDomainSocketPath = source.GetString(ConfigurationKeys.MetricsUnixDomainSocketPath);
-            var partialFlushMinSpans = source.GetInt32(ConfigurationKeys.PartialFlushMinSpans);
+            var dogStatsdPort = new ConfigurationBuilder(source, telemetry).WithKeys(ConfigurationKeys.DogStatsdPort).AsInt32().Get(0);
+            var metricsPipeName = new ConfigurationBuilder(source, telemetry).WithKeys(ConfigurationKeys.MetricsPipeName).AsString().Get();
+            var metricsUnixDomainSocketPath = new ConfigurationBuilder(source, telemetry).WithKeys(ConfigurationKeys.MetricsUnixDomainSocketPath).AsString().Get();
 
             ConfigureTraceTransport(traceAgentUrl, tracesPipeName, agentHost, agentPort, tracesUnixDomainSocketPath);
```

Comment:
We don't capture all the possibilities of issues here. ConfigureTraceTransport can default to different values (if we can't parse the url for instance).

Reply:
I agree, but the logic here is so complex, I'm not sure if we _can_ capture them all. A limitation of the config telemetry is that we need to be able to decide whether a key is valid at the time we retrieve it, which isn't the case with all this extra logic.

I think we _can_ do some extra checks, but I'm a little hesitant to do them in this PR, as they will make things even more complex! 😄 I'll try to address in a subsequent PR

Reply:
If we do anything, it can be done in a different PR for sure. 
I assume an easy solution could be to overwrite the telemetry records _a posteriori_. AgentUri did parse properly to a sting but we then see it's not good, so we override what we have set so far. It's not ideal as you're using a queue to record, but that's maybe one position (and indeed nice to have)